include/nbl/asset/filters/kernels/WeightFunctions.h

```diff
@@ -0,0 +1,308 @@
+};
+
+// Kaiser filter, basically a windowed sinc.
+struct SKaiserFunction
+{
+	constexpr static inline float min_support = -3.f;
+	constexpr static inline float max_support = +3.f;
+	constexpr static inline uint32_t k_smoothness = 3; // TODO(achal): Figure out what this should be
+	// important constant, do not touch, do not tweak
+	static inline constexpr float alpha = 3.f;
+
+	template <int32_t derivative = 0>
+	static inline double weight(float x)
+	{
+		if (x >= min_support && x < max_support)
+		{
+			const float absMinSupport = core::abs(min_support);
+
+			if constexpr (derivative == 0)
+			{
+				const auto PI = core::PI<double>();
+				return core::sinc(x * PI) * core::KaiserWindow(x, alpha, absMinSupport);
+			}
+			else if constexpr (derivative == 1)
+			{
+				const auto PIx = core::PI<double>() * x;
+				double f = core::sinc(PIx);
+				double df = core::PI<double>() * core::d_sinc(PIx);
+				double g = core::KaiserWindow(x, alpha, absMinSupport);
+				double dg = core::d_KaiserWindow(x, alpha, absMinSupport);
+				return df * g + f * dg;
+			}
+			else
+			{
+				static_assert(false, "TODO");
+				return core::nan<double>();
+			}
+		}
+		return 0.0;
+	}
+};
+
+// This is the interface for 1D weight functions that can be used to create a `CFloatingPointSeparableImageFilterKernel`.
+// Current implementations of this interface are:
+// - `CWeightFunction1D`
+// - `CChannelIndependentWeightFunction1D`
+// - `CConvolutionWeightFunction1D`
+template<typename T>
+concept KernelWeightFunction1D = requires(T t, const float x, const uint8_t channel)
+{
+	{ t.weight(x, channel) }	-> std::same_as<double>;
+	{ t.getMinSupport() }		-> std::same_as<float>;
+	{ t.getMaxSupport() }		-> std::same_as<float>;
+	{ t.stretch(x) }			-> std::same_as<void>;
+	{ t.scale(x) }				-> std::same_as<void>;
+	{ t.stretchAndScale(x) }	-> std::same_as<void>;
+};
+
+template <typename Function1D, int32_t derivative = 0>
+class CWeightFunction1D /*final*/ // TODO(achal): Cannot make it final just yet because `weight_function_value_type` inherits from this.
+{
+	static_assert(derivative <= Function1D::k_smoothness);
+public:
+	constexpr static inline uint32_t k_smoothness = Function1D::k_smoothness-derivative;
+	constexpr static inline float k_energy[4] = { 0.f, 0.f, 0.f, 0.f }; // TODO(achal): Implement.
+
+	// Calling: f(x).stretch(2) will obviously give you f(x/2)
+	inline void stretch(const float s)
+	{
+		assert(s != 0.f);
+
+		m_minSupport *= s;
+		m_maxSupport *= s;
+
+		const float rcp_s = 1.f / s;
+
+		m_invStretch *= rcp_s;
+
+		if constexpr (derivative != 0)
+			scale(pow(rcp_s, derivative));
+	}
+
+	inline void scale(const float s)
+	{
+		assert(s != 0.f);
+		m_totalScale *= s;
+	}
+
+	// This method will keep the integral of the weight function constant.
+	inline void stretchAndScale(const float stretchFactor)
+	{
+		stretch(stretchFactor);
+		scale(1.f / stretchFactor);
+	}
+
+	inline double weight(const float x, const uint8_t channel) const
```

Comment:
ok, we should probably stick to Single Responsibility Principle and not bother with worrying about channels in `CWeightFunction1D` and delegate that to a function later down the road

Reply:
done it myself.